### ISSUE TITLE
Error when installing wire-cell-toolkit w/ spack

### ISSUE THREAD

**Issue by ligerlac** [WireCell/wire-cell-spack#9]
When installing wire-cell-toolkit via spack like so:
`spack install wire-cell-toolkit~tbb`
I get the following error in the 'install' phase:
```
==> wire-cell-toolkit: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/bin/bash' '-c' './wcb build --notests install'

1 error found in build log:
     245    [124/256] Compiling gen/src/PerChannelVariation.cxx
     246    [125/256] Compiling gen/src/TrackDepos.cxx
     247    [126/256] Compiling gen/src/NoiseSource.cxx
     248    [127/256] Compiling gen/src/Noise.cxx
     249    [128/256] Compiling gen/src/MultiDuctor.cxx
     250    [129/256] Compiling gen/src/TruthSmearer.cxx
  >> 251    ../gen/src/Noise.cxx:35:21: error: use of overloaded operator '=' is ambiguous (with operand types 'Json::Value' and 'cons
            t size_t' (aka 'const unsigned long'))
     252        cfg["nsamples"] = m_nsamples;
     253        ~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~
     254    /path/spack/local/opt/spack/darwin-monterey-m1/apple-clang-13.1.6/jsoncpp-1.9.5-lkvcop26ytysoe
            7y3likrx36qhfvcdpo/include/json/value.h:352:10: note: candidate function
     255      Value& operator=(const Value& other);
     256             ^
     257    /path/spack/local/opt/spack/darwin-monterey-m1/apple-clang-13.1.6/jsoncpp-1.9.5-lkvcop26ytysoe
            7y3likrx36qhfvcdpo/include/json/value.h:353:10: note: candidate function
```
Not sure if this belongs here or in the wire-cell-toolkit repo.

**Comment by brettviren** [comment]
Hi @ligerlac 

Hmm, it seems that the WCT version being installed via `wire-cell-spack` has a bug.  Indeed, JsonCPP does not know how to convert `size_t` so that error is expected.  I am less sure why my past builds with Spack escaped this bug but we can see it was fixed 7 months ago by @lastgeorge:

https://github.com/WireCell/wire-cell-toolkit/blob/master/gen/src/Noise.cxx#L35

It looks like that fix came in just after `0.24.1` that current `wire-cell-spack` builds.  You can add newer releases in your local Spack repo in this way: 

```
$ spack checksum wire-cell-toolkit 0.24.2 0.25.1 0.25.2
==> Found 3 versions of wire-cell-toolkit:
  
  0.25.2  https://github.com/WireCell/wire-cell-toolkit/archive/refs/tags/0.25.2.tar.gz
  0.25.1  https://github.com/WireCell/wire-cell-toolkit/archive/refs/tags/0.25.1.tar.gz
  0.24.2  https://github.com/WireCell/wire-cell-toolkit/archive/refs/tags/0.24.2.tar.gz

==> Fetching https://github.com/WireCell/wire-cell-toolkit/archive/refs/tags/0.25.2.tar.gz
==> Fetching https://github.com/WireCell/wire-cell-toolkit/archive/refs/tags/0.25.1.tar.gz
==> Fetching https://github.com/WireCell/wire-cell-toolkit/archive/refs/tags/0.24.2.tar.gz

    version("0.25.2", sha256="aac849da29092d3397f0130f45c499703496188ad74bd347a3bd154fe23f245d")
    version("0.25.1", sha256="c236758a70fd459824f2f769b1e4e146e8f75965833a5aa8a81f0cf08044f9d6")
    version("0.24.2", sha256="ee8fe70f8ead7b7456bbcd791d8bf1a1cf22c6674df503fb9f93c1d33bfd2a1a")

```

These `version()` lines get pasted above the existing ones in `wire-cell-toolkit/plugin.py`.  You can then redo `spack install wire-cell-toolkit` and that should default to the first `version()` found.

If you want to try this and if you have success feel free to make a PR with the updated `package.py`.  If you have troubles let me know and I'll do it.

**Comment by ligerlac** [comment]
Hey @brettviren,

I have created the PR here: https://github.com/WireCell/wire-cell-spack/pull/10

Is there any reason why a certain tag should not appear in `package.py`? If no, what do you think of automating the process? I think this could be achieved with a github Actions workflow.

**Comment by brettviren** [comment]
Hi @ligerlac 

> Is there any reason why a certain tag should not appear in `package.py`? If no, what do you think of automating the process? I think this could be achieved with a github Actions workflow.

I think in general adding every tagged release to the `package.py` is fine and automating it would be very useful. In fact, WCT ecosystem would generally benefit from more CI automation but so far we have lacked the expertise/effort to make improvements there.

Automatically prepending `version()` lines for each release of WCT would need to be arranged with some assumptions about post-hoc changes:

- We would likely purge some `version()` lines either because a recent release had issues and should be avoided or simply to keep the list a "reasonable" length.  If it could be automated to only keep the largest version of any release (eg purge `X.YY.1` when `X.YY.2` came out) that would likely implement most of the kinds of pruning we might otherwise do manually.
- Over history we will add new or remove old dependencies at a certain version.  This happens lower in `package.py` below the `version()` lines and likely won't matter to any automated updating, but I just mention it.

BTW, WCT has a version convention that `X.YY.0` starts a "release branch".  However, we often forget to make the `.0` an explicit branch.  When I find a bug in a release and find that there is only a `.0` tag and no branch then I must remind myself how to do the git dance to manually add the branch.  Not hard, but I never remember it so it adds friction.  If CI could automatically make a `.0` branch to go with any newly discovered `.0` tag, that would be very welcome!  (assuming it's easy enough to set that up).

**Comment by brettviren** [comment]
Thanks for the PR.  I close this.  

BTW, GH will auto close an issue when commits are merged with commit text like: "fix #9" or "fixes #9".  (Apparently though, it doesn't do this with comments on the PR, at least not after the merge which I just tried).
